client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java

```diff
@@ -0,0 +1,306 @@
+
+        public ConnectionReader(String name, InputStream in, ReplyProcessor callback,
+                                AppendBatchSizeTracker batchSizeTracker) {
+            this.name = name;
+            this.in = in;
+            this.callback = callback;
+            this.thread = ExecutorServiceHelpers.newScheduledThreadPool(1, "Reading from " + name);
+            this.batchSizeTracker = batchSizeTracker;
+        }
+        
+        public void start() {
+            thread.submit(() -> {
+                IoBuffer buffer = new IoBuffer();
+                while (!stop.get()) {
+                    try {
+                        WireCommand command = readCommand(in, buffer);
+                        if (command instanceof WireCommands.DataAppended) {
+                            WireCommands.DataAppended dataAppended = (WireCommands.DataAppended) command;
+                            batchSizeTracker.recordAck(dataAppended.getEventNumber());
+                        }
+                        try {
+                            callback.process((Reply) command);
+                        } catch (Exception e) {
+                            callback.processingFailure(e);
+                        }
+                    } catch (SocketException e) {
+                        if (e.getMessage().equals("Socket closed")) {
+                            log.info("Closing TcpConnection.Reader because socket is closed.");
+                        } else {
+                            log.warn("Error in reading from socket.", e);
+                        }
+                        stop();
+                    } catch (EOFException e) {
+                        log.info("Closing TcpConnection.Reader because end of input readched.");
+                        stop();
+                    } catch (Exception e) {
+                        log.warn("Error processing data from from server " + name, e);
+                        stop();
+                    }
+                }
+            });
+        }
+
+        @VisibleForTesting
+        static WireCommand readCommand(InputStream in, IoBuffer buffer) throws IOException {
+            ByteBuf header = buffer.getBuffOfSize(in, 8);
+
+            int t = header.getInt(0);
+            WireCommandType type = WireCommands.getType(t);
+            if (type == null) {
+                throw new InvalidMessageException("Unknown wire command: " + t);
+            }
+
+            int length = header.getInt(4);
+            if (length < 0 || length > WireCommands.MAX_WIRECOMMAND_SIZE) {
+                throw new InvalidMessageException("Event of invalid length: " + length);
+            }
+
+            ByteBuf payload = buffer.getBuffOfSize(in, length);
+            
+            return type.readFrom(new EnhancedByteBufInputStream(payload), length);
+        }
+        
+        public void stop() {
+            stop.set(true);
+            thread.shutdown();
```

Comment:
`shutdownNow` or `ExecutorServiceHelpers.shutdown`; otherwise there's no guarantee that its threads are properly interrupted.

Reply:
shall we wait for the threadPool to terminate execution ?

Reply:
Removed the pool